data/bookkeeping/block_test.go

```diff
@@ -461,3 +462,72 @@ func TestInitialRewardsRateCalculation(t *testing.T) {
 	consensusParams.InitialRewardsRateCalculation = true
 	require.True(t, runTest())
 }
+
+func performRewardsRateCalculation(
+	t *testing.T, consensusParams config.ConsensusParams,
+	curRewardsState RewardsState,
+	incentivePoolBalance uint64, totalRewardUnits uint64, startingRound uint64) {
+	require.GreaterOrEqual(t, incentivePoolBalance, consensusParams.MinBalance)
+
+	for rnd := startingRound; rnd < startingRound+uint64(consensusParams.RewardsRateRefreshInterval)*3; rnd++ {
+		nextRewardState := curRewardsState.NextRewardsState(basics.Round(rnd), consensusParams, basics.MicroAlgos{Raw: incentivePoolBalance}, totalRewardUnits)
+		// adjust the incentive pool balance
+		var ot basics.OverflowTracker
+
+		// get number of rewards per unit
+		rewardsPerUnit := ot.Sub(nextRewardState.RewardsLevel, curRewardsState.RewardsLevel)
+		require.False(t, ot.Overflowed)
+
+		// subtract the total dispersed funds from the pool balance
+		incentivePoolBalance = ot.Sub(incentivePoolBalance, ot.Mul(totalRewardUnits, rewardsPerUnit))
+		require.False(t, ot.Overflowed)
+
+		require.GreaterOrEqual(t, incentivePoolBalance, consensusParams.MinBalance, rnd)
+		// prepare for the next iteration
+		curRewardsState = nextRewardState
+	}
+}
+
+func TestNextRewardsRateWithFix(t *testing.T) {
+	partitiontest.PartitionTest(t)
+
+	params, ok := config.Consensus[protocol.ConsensusCurrentVersion]
+	require.True(t, ok)
+	params.RewardsCalculationFix = true
```

Comment:
The test is "almost there". You'll need to confirm expected failure without the fix, and expected succeess with the fix.

Reply:
It would also be a good idea to add expected failure after the fix, providing invalid input.

Reply:
Done (1). Why test on invalid input?

Reply:
testing it against invalid input ensures that the method does behave in the way we expect it to. I realize that there could be multitude of invalid inputs, and ideally, we need to analyze the codebase to ensure all the code-paths are taken and validated individually.